util/chunk/chunk.go

```diff
@@ -112,12 +159,68 @@ func (c *Chunk) SetNumVirtualRows(numVirtualRows int) {
 // Reset resets the chunk, so the memory it allocated can be reused.
 // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
 func (c *Chunk) Reset() {
-	for _, c := range c.columns {
-		c.reset()
+	c.GrowAndReset(c.capacity)
+}
+
+// GrowAndReset resets the chunk and grow capacity if need.
+// TODO: this method will be used in continue PR.
+func (c *Chunk) GrowAndReset(capLimit int) {
+	if c.columns == nil {
+		return
+	}
+	requireRenewCol, newCap := calculateResetCap(c, capLimit)
+	if !requireRenewCol {
+		for _, col := range c.columns {
+			col.reset()
+		}
+		c.numVirtualRows = 0
+		return
 	}
+	c.capacity = newCap
+	c.columns = renewColumns(c.columns, newCap)
 	c.numVirtualRows = 0
 }
 
+// calculateResetCap checks and return whether need scale up and suitable capacity execution.
+//  It works like:
+//      keep old value if capacity == capLimit or chk isn't full, and no need renew columns.
+//      old value x 2  if chk is full, and need renew columns.
+func calculateResetCap(c *Chunk, capLimit int) (requireRenewCol bool, newCap int) {
+	if c.capacity == capLimit || c.NumRows() < c.capacity {
+		return false, c.capacity
+	}
+	return true, mathutil.Min(c.capacity*2, capLimit)
+}
+
+// calculateResetCap checks and return suitable capacity in next execution.
+//  It works like:
+//      using next power of 2 for renew if currSize < cap
+//      using capacity limit if full but capacity already meet capLimit
+//      using oldcap x 2 if chk is full but less than capLimit
+func calculateRenewCap(c *Chunk, capLimit int) int {
```

Comment:
Maybe we only need `calculateRenewCap`, and `calculateResetCap` can be removed?

Reply:
this two method has some different

calculateResetCap: if previous return chk not full, will keep not grow.
calculateRenewCap: if previous returned chk  is not full, will have chance give a smaller capacity then previous. 

as https://github.com/pingcap/tidb/pull/7473#discussion_r212549032 advise I divide them into two.